### PR TITLE
fix(google-provider): differences between web and desktop in google oauth

### LINKED ISSUE #2905: google provider registration (web)
- Register ouath client with google (web application type) image

- Please also set "Authorized JavaScript origins" and "Authorized redirect URIs" to the same values that are registered with Dropbox

### PR DESCRIPTION
**Pull request by mroz22**
If I am correct, google [does not allow](https://developers.google.com/identity/protocols/oauth2#clientside
) long-lived tokens on client side javascript in browser (without backend) 

It is actually  possible to use client_id for application of type iOS (which we use for [desktop](https://developers.google.com/identity/protocols/oauth2#clientside
)) for web on localhost, but not in production (#2528).You can try it on current develop branch after flipping google feature flag to true. 

So the only way around this (and please correct me if I am wrong) is to support implicit grant for google/web and authorization code flow for google/desktop.

changes:
- metadata providers now have "supportsTokenRenewal" field and select appropriate authorization flow 
- tests now run againts both cloud providers

todo:
- [ ] official google/web clientId

___

close #2905

### CODE REVIEW

**Comment by mroz22** [comment]
Ok, mosty done here. I'd say it is ready to review, it just needs to add prober app id which is the last todo to do.

File: packages/suite/src/actions/suite/constants/metadataConstants.ts [hunk 1]
```diff
@@ -26,9 +26,14 @@ export const AUTH_WINDOW_WIDTH = 600;
...
 
-// todo: still thinking about this google token. this one is for iOS (which is recommended token type to be used in electron). It works
-// for both electron and web. But we also may have web token which works only for web. Web has restriction on redirect urls, iOs does not.
-// we use PKCE for both. Is there any additional security in using token for web? Hmmm...
-export const GOOGLE_CLIENT_ID =
+// google does not allow authorization code flow (with long lived tokens) for web. that is the reason why 2 clientIds are used
+export const GOOGLE_CLIENT_ID_DESKTOP =
     '705190185912-q1usa46qtt21mbtldki8juhn78a0v0ma.apps.googleusercontent.com';
+export const GOOGLE_CLIENT_ID_WEB =
+    '842348096891-efhc485636d5t09klvrve0pi4njhq3l8.apps.googleusercontent.com';
```
Comment (mroz22) [inline packages/suite/src/actions/suite/constants/metadataConstants.ts]: todo

File: packages/suite-data/files/oauth/oauth_receiver.html [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
-    </script>
-  </body>
-</html>
\ No newline at end of file
+    <head></head>
+    <body>
+        <script>
+            // google implicit flow - hash
+            // dropbox authorization code flow - search
+            var message = window.location.hash || window.location.search;
```
Comment (szymonlesisz) [inline packages/suite-data/files/oauth/oauth_receiver.html]: use both in message (search and hash)

File: packages/suite/src/services/google.ts [hunk 3]
```diff
@@ -95,45 +95,82 @@ class Client {
         this.token = token;
         this.nameIdMap = {};
         this.oauth2Client = new OAuth2Client({
-            clientId: METADATA.GOOGLE_CLIENT_ID,
+            clientId:
+                process.env.SUITE_TYPE === 'web'
```
Comment (szymonlesisz) [inline packages/suite/src/services/google.ts]: isWeb utility

File: packages/suite/src/services/google.ts [hunk 4]
```diff
@@ -287,6 +324,7 @@ class Client {
             const query = new URLSearchParams(apiParams.query as Record<string, string>).toString();
             url += `?${query}`;
         }
+
         const accessToken = await this.oauth2Client.getAccessToken();
```
Comment (szymonlesisz) [inline packages/suite/src/services/google.ts:326]: ? token validation discussion

File: packages/suite/src/utils/suite/oauth.ts [hunk 5]
```diff
@@ -18,10 +17,11 @@ export const getOauthReceiverUrl = () => {
 /**
  * Handle extraction of authorization code from Oauth2 protocol
  */
-export const getOauthCode = (url: string) => {
+export const extractCredentialsFromAuthorizationFlow = (url: string) => {
     const originalParams = urlHashParams(url);
 
-    const dfd: Deferred<string> = createDeferred();
+    // eslint-disable-next-line camelcase
+    const dfd: Deferred<{ code?: string; access_token?: string }> = createDeferred();
```
Comment (szymonlesisz) [inline packages/suite/src/utils/suite/oauth.ts]: ? types union discussion

**Review by szymonlesisz — CHANGES_REQUESTED** [review]

**Comment by mroz22** [comment]
hmm one more todo: 
- [x] providers that do not support token renewal should be periodically fetching data (and thus validate token under the hood). If it turns out that token is no logner valid, such provider should not react with displaying error but simply disconnect provider while keeping metadata in place? So that user will see all metadata even after provider has disconnected. Just won't be able to edit them without manual reauthentication.

File: packages/suite/src/services/google.ts (hunk 4 above)
Comment (mroz22) [reply packages/suite/src/services/google.ts:326]: Should be ok! From jsdoc definition: 
```
 /**
     * Get a non-expired access token, after refreshing if necessary
     *
     * @param callback Callback to call with the access token
     */
```

**Comment by mroz22** [comment]
Review requests addressed in 20d1af3010a97dfe6e374fd9784960b2a1705a4d

**Review by szymonlesisz — APPROVED** [review]

**Pull request merged**